website/blog/how-bazel-5-0-makes-your-builds-faster.md

```diff
@@ -0,0 +1,133 @@
+then this caused significant network overhead.
+
+Bazel 5.0 addresses these issues.
+If you have a project that shares a lot of inputs between actions
+(e.g. C, C++, Objective-C, and Node.js),
+and you use BuildBuddy's Remote Execution,
+then Bazel 5.0 should improve your build times.
+
+[^1]:
+    Specifically,
+    Bazel would make a `FindMissingBlobs` that contained every input for an action,
+    every time.
+    Now Bazel will only include the unique inputs it hasn't already uploaded before.
+
+## Faster action cache checking
+
+In a similar vein as the above mentioned issues,
+Bazel can be unnecessarily slow when checking if an action with a large number of inputs is cached.
+This is because Bazel needs to build a [Merkle tree][merkle_tree] from an action's inputs,
+which can be expensive if there are many.
+
+Starting in Bazel 5.0,
+you can use the `--experimental_remote_merkle_tree_cache` and `--experimental_remote_merkle_tree_cache_size` flags to cache the nodes of these merkle trees.
+If you have a project that shares a lot of it's inputs between actions
+(e.g. C, C++, Objective-C, and Node.js),
+and you use BuildBuddy's Remote Cache or Remote Execution,
+then these flags should improve your build times.
+
+[merkle_tree]: https://en.wikipedia.org/wiki/Merkle_tree
+
+## Asynchronous cache uploading
+
+When an action is run locally,
+and Bazel uploads the action's outputs to a remote build cache,
+it waits for that upload to complete before it executes dependent actions.
+This is needed for dependent actions that are executed remotely,
+but it's unnecessary if they are executed locally.
+These delays can cause builds that use BuildBuddy's Remote Cache to sometimes be slower than if they didn't use the cache.
+
+Starting in Bazel 5.0,
+you can use the `--experimental_remote_cache_async` flag to have Bazel perform these uploads asynchronously.
+Uploads still need to complete before a build finishes,
+but now Bazel can speed ahead with local execution regardless of upload speed.
+Network utilization is also improved,
+as uploads can queue up quicker.
+If you have a project that uploads to BuildBuddy's Remote Cache,
+then this flag should improve your build times.
+
+## Compressed uploads and downloads
+
+Bazel 5.0 added the `--experimental_remote_cache_compression` flag,
+which causes Bazel to compresses and decompress artifact uploads and downloads with the [zstd algorithm][zstd].
+In our testing this reduces the average number of bytes transferred by 60-70%.
+
+There is some overhead involved in this,
+both on the Bazel side and the cache side,
+so it won't make every build for every project faster.[^2]
+That said,
+if network bandwidth is a bottleneck in your build times,
+this flag could make those builds significantly faster.
+
+[zstd]: https://en.wikipedia.org/wiki/Zstandard
+
+[^2]:
+    For now.
+    Eventually we might store artifacts in their compressed state,
+    removing all overhead on the cache side.
```

Comment:
For more context, there will still be some overhead on the cache side because we'll need decompressive transcoding on uploads in order to verify their checksum. Also, for RBE inputs/outputs we'd need both compressive and decompressive transcoding either on the cache or on the executor (most likely, it makes sense to do it on the cache, since CPU resources are more under contention on the executor). So for folks using RBE, it's a bit unclear at the moment whether storing artifacts in their compressed state would be a net benefit.

Reply:
:+1:

I'll tweak the wording a bit.